File: plc4j/spi/src/main/java/org/apache/plc4x/java/spi/generation/BufferCommons.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.plc4x.java.spi.generation;

import org.apache.commons.lang3.StringUtils;

import java.util.Optional;
import java.util.stream.Stream;

public interface BufferCommons {
    String rwDataTypeKey = "dataType";
    String rwBitLengthKey = "bitLength";
    String rwStringRepresentationKey = "stringRepresentation";
    String rwBitKey = "bit";
    String rwByteKey = "byte";
    String rwUintKey = "uint";
    String rwIntKey = "int";
    String rwFloatKey = "float";
    String rwStringKey = "string";
    String rwEncodingKey = "encoding";
    String rwIsListKey = "isList";

    default String sanitizeLogicalName(String logicalName) {
        if (StringUtils.isBlank(logicalName)) {
            return "value";
        }
        return logicalName;
    }

    default boolean isToBeRenderedAsList(WithReaderArgs... readerArgs) {
        return isToBeRenderedAsList(Stream.of(readerArgs).map(WithReaderWriterArgs.class::cast).toArray(WithReaderWriterArgs[]::new));
    }

    default boolean isToBeRenderedAsList(WithWriterArgs... writerArgs) {
        return isToBeRenderedAsList(Stream.of(writerArgs).map(WithReaderWriterArgs.class::cast).toArray(WithReaderWriterArgs[]::new));
    }

    default boolean isToBeRenderedAsList(WithReaderWriterArgs... readerWriterArgs) {
        for (WithReaderWriterArgs arg : readerWriterArgs) {
            if (arg instanceof withRenderAsList) {
                return ((withRenderAsList) arg).renderAsList();
            }
        }
        return false;
    }

    default Optional<String> extractAdditionalStringRepresentation(WithReaderArgs... readerArgs) {
        return extractAdditionalStringRepresentation(Stream.of(readerArgs).map(WithReaderWriterArgs.class::cast).toArray(WithReaderWriterArgs[]::new));
    }

    default Optional<String> extractAdditionalStringRepresentation(WithWriterArgs... writerArgs) {
        return extractAdditionalStringRepresentation(Stream.of(writerArgs).map(WithReaderWriterArgs.class::cast).toArray(WithReaderWriterArgs[]::new));
    }

    default Optional<String> extractAdditionalStringRepresentation(WithReaderWriterArgs... readerWriterArgs) {
        for (WithReaderWriterArgs arg : readerWriterArgs) {
            if (arg instanceof withAdditionalStringRepresentation) {
                return Optional.of(((withAdditionalStringRepresentation) arg).stringRepresentation());
            }
        }
        return Optional.empty();
    }

    default Optional<String> extractEncoding(WithReaderArgs... readerArgs) {
        return extractEncoding(Stream.of(readerArgs).map(WithReaderWriterArgs.class::cast).toArray(WithReaderWriterArgs[]::new));
    }

    default Optional<String> extractEncoding(WithWriterArgs... writerArgs) {
        return extractEncoding(Stream.of(writerArgs).map(WithReaderWriterArgs.class::cast).toArray(WithReaderWriterArgs[]::new));
    }

    default Optional<String> extractEncoding(WithReaderWriterArgs... readerWriterArgs) {
        for (WithReaderWriterArgs arg : readerWriterArgs) {
            if (arg instanceof withEncoding) {
                return Optional.of(((withEncoding) arg).encoding());
            }
        }
        return Optional.empty();
    }

}

File: plc4go/internal/plc4go/spi/utils/bufferCommons.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package utils

import "container/list"

const (
	rwDataTypeKey             = "dataType"
	rwBitLengthKey            = "bitLength"
	rwStringRepresentationKey = "stringRepresentation"
	rwBitKey                  = "bit"
	rwByteKey                 = "byte"
	rwUintKey                 = "uint"
	rwIntKey                  = "int"
	rwFloatKey                = "float"
	rwStringKey               = "string"
	rwEncodingKey             = "encoding"
	rwIsListKey               = "isList"
)

type bufferCommons struct {
}

func (b bufferCommons) sanitizeLogicalName(logicalName string) string {
	if logicalName == "" {
		return "value"
	}
	return logicalName
}

func (b bufferCommons) isToBeRenderedAsList(readerWriterArgs ...WithReaderWriterArgs) bool {
	for _, arg := range readerWriterArgs {
		if !arg.isWriterArgs() && !arg.isReaderArgs() {
			panic("not a reader or writer arg")
		}
		switch arg.(type) {
		case withRenderAsList:
			return arg.(withRenderAsList).renderAsList
		}
	}
	return false
}

func (b bufferCommons) extractAdditionalStringRepresentation(readerWriterArgs ...WithReaderWriterArgs) string {
	for _, arg := range readerWriterArgs {
		if !arg.isWriterArgs() && !arg.isReaderArgs() {
			panic("not a reader or writer arg")
		}
		switch arg.(type) {
		case withAdditionalStringRepresentation:
			return arg.(withAdditionalStringRepresentation).stringRepresentation
		}
	}
	return ""
}

type stack struct {
	list.List
}

func (s *stack) Push(value interface{}) interface{} {
	s.PushBack(value)
	return value
}

func (s *stack) Pop() interface{} {
	if s.Len() <= 0 {
		return nil
	}
	element := s.Back()
	if element == nil {
		return nil
	}
	s.Remove(element)
	return element.Value
}

func (s *stack) Peek() interface{} {
	back := s.Back()
	if back == nil {
		return nil
	}
	return back.Value
}

func (s stack) Empty() bool {
	return s.Len() == 0
}
